modules/imgproc/src/contours_approx.cpp

```diff
@@ -0,0 +1,353 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#include "opencv2/core/base.hpp"
+#include "opencv2/core/types.hpp"
+#include "opencv2/imgproc.hpp"
+#include <vector>
+
+using namespace std;
+using namespace cv;
+
+namespace {
+
+struct ApproxItem
+{
+    Point pt;
+    int k; // support region
+    int s; // 1-curvature
+    bool removed;
+    ApproxItem() : k(0), s(0), removed(false) {}
+    ApproxItem(Point pt_, int s_) : pt(pt_), k(0), s(s_), removed(false) {}
+};
+
+static const schar abs_diff[16] = { 1, 2, 3, 4, 3, 2, 1, 0, 1, 2, 3, 4, 3, 2, 1 };
+static const Point chainCodeDeltas[8] =
+    { {1, 0}, {1, -1}, {0, -1}, {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1} };
+
+// Pass 0.
+// Restores all the digital curve points from the chain code.
+// Removes the points (from the resultant polygon)
+// that have zero 1-curvature
+static vector<ApproxItem> pass_0(const vector<schar> & chain, Point pt, bool isApprox, bool isFull)
+{
+    vector<ApproxItem> res;
+    const size_t len = chain.size();
+    for (size_t i = 0; i < len; ++i)
+    {
+        const schar prev = (i == 0) ? chain[len - 1] : chain[i - 1];
+        const schar cur = chain[i];
+        const schar s = abs_diff[cur - prev + 7];
+        if ((!isApprox && (isFull || s != 0)) || isApprox)
+        {
+            res.push_back(ApproxItem(pt, s));
+            if (s == 0)
+                (res.end() - 1)->removed = true;
+        }
+        pt += chainCodeDeltas[cur];
+    }
+    return res;
+}
+
+static vector<Point> gatherPoints(const vector<ApproxItem> & ares)
+{
+    vector<Point> res;
+    for (const ApproxItem & item : ares)
+    {
+        if (item.removed)
+            continue;
+        res.push_back(item.pt);
+    }
+    return res;
+}
+
+static int calc_support(const vector<ApproxItem> & ares, int i)
+{
+    int k;
+    int d_num = 0;
+    int l = 0;
+    const int len = (int)ares.size();
+    /* determine support region */
+    for( k = 1;; k++ )
+    {
+        CV_Assert( k <= len );
```

Comment:
What if add the check to for loop?

Reply:
This was in the original code and, since this function isn't covered by tests, I tried to preserve this original behavior. As I understand, `k` should't exceed `len` (even `len/2` maybe) and if this happens then it means that input array was not valid, so exception might be more appropriate than continuing.